source_middleware.go

```diff
@@ -409,3 +382,142 @@ func (s *sourceWithSchemaExtraction) encodeWithSchema(sch schema.Schema, data an
+	return "sdk.schema.context.name"
+}
+
+// SourceWithSchemaContext is a middleware that makes it possible to configure
+// the schema context for records read by a source.
+type SourceWithSchemaContext struct {
+	Config SourceWithSchemaContextConfig
+}
+
+// Wrap a Source into the schema middleware. It will apply default configuration
+// values if they are not explicitly set.
+func (s *SourceWithSchemaContext) Wrap(impl Source) Source {
+	if s.Config.Enable == nil {
+		s.Config.Enable = ptr(true)
+	}
+
+	return &sourceWithSchemaContext{
+		Source: impl,
+		mwCfg:  s.Config,
+	}
+}
+
+type sourceWithSchemaContext struct {
+	Source
+	// mwCfg is the default middleware config
+	mwCfg SourceWithSchemaContextConfig
+
+	useContext  bool
+	contextName string
+}
+
+func (s *sourceWithSchemaContext) Parameters() config.Parameters {
+	return mergeParameters(s.Source.Parameters(), s.mwCfg.parameters())
+}
+
+func (s *sourceWithSchemaContext) Configure(ctx context.Context, cfg config.Config) error {
+	s.useContext = *s.mwCfg.Enable
+	if useStr, ok := cfg[s.mwCfg.EnableParameterName()]; ok {
+		use, err := strconv.ParseBool(useStr)
+		if err != nil {
+			return fmt.Errorf("could not parse `%v`, input %v: %w", s.mwCfg.EnableParameterName(), useStr, err)
+		}
+		s.useContext = use
+	}
+
+	if s.useContext {
+		// the order of precedence (least to most) is:
+		// 1. connector ID (if no context name is configured anywhere)
+		// 2. default middleware config
+		// 3. user config
+		s.contextName = internal.ConnectorIDFromContext(ctx)
+		if s.mwCfg.Name != nil {
+			s.contextName = *s.mwCfg.Name
+		}
+		if ctxName, ok := cfg[s.mwCfg.NameParameterName()]; ok {
+			s.contextName = ctxName
+		}
+	}
+
+	return s.Source.Configure(sdkschema.WithSchemaContextName(ctx, s.contextName), cfg)
+}
+
+func (s *sourceWithSchemaContext) Open(ctx context.Context, pos opencdc.Position) error {
+	return s.Source.Open(sdkschema.WithSchemaContextName(ctx, s.contextName), pos)
+}
+
+func (s *sourceWithSchemaContext) Read(ctx context.Context) (opencdc.Record, error) {
+	return s.Source.Read(sdkschema.WithSchemaContextName(ctx, s.contextName))
+}
+
+func (s *sourceWithSchemaContext) Teardown(ctx context.Context) error {
+	return s.Source.Teardown(sdkschema.WithSchemaContextName(ctx, s.contextName))
+}
+
+func (s *sourceWithSchemaContext) LifecycleOnCreated(ctx context.Context, config config.Config) error {
+	return s.Source.LifecycleOnCreated(sdkschema.WithSchemaContextName(ctx, s.contextName), config)
+}
+
+func (s *sourceWithSchemaContext) LifecycleOnUpdated(ctx context.Context, configBefore, configAfter config.Config) error {
+	return s.Source.LifecycleOnUpdated(sdkschema.WithSchemaContextName(ctx, s.contextName), configBefore, configAfter)
+}
+
+func (s *sourceWithSchemaContext) LifecycleOnDeleted(ctx context.Context, config config.Config) error {
+	return s.Source.LifecycleOnDeleted(sdkschema.WithSchemaContextName(ctx, s.contextName), config)
+}
+
+func ptr[T any](t T) *T {
```

Comment:
👍 I feel like we could add a utility like this somewhere into conduit-common, every now and then we need it. Just a thought, no need to do it now.

Reply:
I was thinking it's maybe too simple to be provided as a library function:), but I can definitely add that.